src/dataflow-types/src/reconciliation/command.rs

```diff
@@ -0,0 +1,528 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+//! Functionality to reconcile commands between a COMPUTE controller and a COMPUTE instance.
+//!
+//! The [ComputeCommandReconcile] and [StorageCommandReconcile] structs implement [GenericClient]
+//! the controller to reconnect after restarts. It maintains enough state to
+//! get a newly connected instance up-to-date and matches existing installed objects with
+//! what the controller wants to provide.
+//!
+//! [ComputeCommandReconcile] is designed to live in a COMPUTE instance and liberally uses `assert` to
+//! validate the correctness of commands. It is not intended to be part of a COMPUTE coordinator as
+//! it might have correctness issues.
+//!
+//! The contract between this and the COMPUTE controller is that identifiers are not re-used and
+//! describe the same object after restarts. Failure to do so will result in undefined behavior.
+//!
+//! The reconciliation presents to a restarted COMPUTE coordinator as if the COMPUTE instance was
+//! restarted as well. It responds with the expected replies after a `CreateInstance` command and
+//! brings the controller  up-to-date by notifying it about the current upper frontiers.
+//!
+//! Controllers should ignore all responses received before `CreateInstance` as those were intended
+//! for the previous instance. The implementation currently does not distinguish between buffering
+//! messages for a disconnected controller and talking to a live controller.
+//!
+//! [StorageCommandReconcile] is designed to live in a STORAGE instance.
+
+use crate::client::{
+    Command, ComputeCommand, ComputeInstanceId, ComputeResponse, GenericClient, Response,
+    StorageCommand, StorageResponse, TimestampBindingFeedback,
+};
+use crate::sources::SourceConnector;
+use crate::{DataflowDescription, Plan};
+use async_trait::async_trait;
+use mz_expr::GlobalId;
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, HashSet, VecDeque};
+use timely::progress::frontier::MutableAntichain;
+use timely::progress::ChangeBatch;
+use tracing::warn;
+
+/// Reconcile commands targeted at a COMPUTE instance.
+///
+/// See the module-level documentation for details.
+#[derive(Debug)]
+pub struct ComputeCommandReconcile<T, C> {
```

Comment:
It feels like much of this struct is redundant with `ComputeState`. Does that seem right, and should it eventually be fused with that struct, once we like how it behaves?

Reply:
We could probably merge the two once we have the confidence it works well. Currently, command reconciliation sits on the `dataflowd` ingress, and merging its state with `ComputeState` would mean it needs to be moved closer to the worker, or after `Partitioned`. I think it's doable and might simplify some of the tracking done here.

Reply:
That makes sense that it is one-per-`dataflowd`. I retract my eagerness!